pkg/k8sclient/client.go

```diff
@@ -0,0 +1,177 @@
+}
+
+var _ Interface = &client{}
+
+type client struct {
+	inner kubernetes.Interface
+}
+
+// NewForConfig creates a client for the specified rest.Config config.
+// It is a wrapper around kubernetes.NewForConfig, but ensures a short
+// timeout.
+func NewForConfig(config *rest.Config) (Interface, error) {
+	// Set a lower timeout, to work around
+	// https://github.com/kubernetes/client-go/issues/374 We
+	// trigger a timeout error and then to recover we need to
+	// reset any existing connections
+	if config.Timeout < maxTimeout {
+		config.Timeout = maxTimeout
+	}
+
+	c, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return &client{inner: c}, nil
+}
+
+func (c *client) RawClient() kubernetes.Interface {
+	return c.inner
+}
+
+func (c *client) ListNodes(ctx context.Context) (*corev1.NodeList, error) {
+	client := c.RawClient()
+	nodes, err := client.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+	c.handleError(err)
+	return nodes, err
+}
+
+func (c *client) DeleteNode(ctx context.Context, nodeName string) error {
+	client := c.RawClient()
+	err := client.CoreV1().Nodes().Delete(ctx, nodeName, metav1.DeleteOptions{})
+	c.handleError(err)
+	return err
+}
+
+// TaintNode applies the taint to the specified node
+func TaintNode(ctx context.Context, k8sClient Interface, node *corev1.Node, taint corev1.Taint) error {
+	oldData, err := json.Marshal(node)
+	if err != nil {
+		return err
+	}
+
+	node.Spec.Taints = append(node.Spec.Taints, taint)
+
+	newData, err := json.Marshal(node)
+	if err != nil {
+		return err
+	}
+
+	patchBytes, err := strategicpatch.CreateTwoWayMergePatch(oldData, newData, node)
+	if err != nil {
+		return err
+	}
+
+	_, err = k8sClient.RawClient().CoreV1().Nodes().Patch(ctx, node.Name, types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{})
+	if c, ok := k8sClient.(*client); ok {
+		c.handleError(err)
+	} else {
+		klog.Warningf("unexpected type for client %T", c)
+	}
+	return err
+}
+
+// handleError is called with all potential errors.
+// When it observes a timeout error, it closes all open / idle connections.
+func (c *client) handleError(err error) {
+	if err == nil {
+		return
+	}
+
+	isTimeout := false
+	s := err.Error()
+	// TODO: move this to the chained errors framework when that's available
+	if strings.Contains(s, "Client.Timeout exceeded while awaiting headers") {
+		isTimeout = true
+	}
+
+	if isTimeout {
+		restClientInterface := c.inner.CoreV1().RESTClient()
+		restClient, ok := restClientInterface.(*rest.RESTClient)
+		if !ok {
+			klog.Warningf("client timed out, but rest client was not of expected type, was %T", restClientInterface)
+			return
+		}
+
+		httpTransport := findHTTPTransport(restClient.Client.Transport)
+		if httpTransport == nil {
+			klog.Warningf("client timed out, but http transport was not of expected type, was %T", restClient.Client.Transport)
+			return
+		}
+		httpTransport.CloseIdleConnections()
```

Comment:
@justinsb How would this make sure that the connection over which the request just timed out is idle and hence would be closed? Is there a possibility that the connection is picked up by another request and is in use while this is getting executed? // @johngmyers

Reply:
If the connection got picked up by another request, that's fine. When the connection does go idle the last request to complete will close it.

Reply:
@johngmyers Sure, this may work in this particular case. Would the same approach work in general for some other high throughput clients suffering from connection staleness? Sorry to hijack this thread, but my issue is similar but related to a different system, and was wondering if the same approach can be applied there.